app/controllers/storage_controller.rb

```diff
@@ -121,78 +122,44 @@ def show(record = nil)
     @lastaction = "show"
   end
 
+  def specific_buttons(pressed)
+    case pressed
+    when "custom_button"                  then custom_buttons
+    when "host_analyze_check_compliance"  then handle_analyze_check_compliance_hosts
+    when "host_check_compliance"          then handle_analyze_check_compliance_hosts
+    when "host_delete"                    then deletehosts
+    when "host_edit"                      then edit_record
+    when "host_protect"                   then assign_policies(Host)
+    when "host_refresh"                   then refreshhosts
+    when "host_scan"                      then scanhosts
+    when "host_tag"                       then tag(Host)
+    when "storage_delete"                 then deletestorages
+    when "storage_refresh"                then refreshstorage
+    when "storage_scan"                   then scanstorage
+    when "storage_tag"                    then tag(Storage)
+    end
 
-  # handle buttons pressed on the button bar
-  def button
-    @edit = session[:edit]                                  # Restore @edit for adv search box
-    params[:display] = @display if %w(all_vms vms hosts).include?(@display) # Were we displaying vms or hosts
+    handle_host_power_buttons(pressed)
 
-    if params[:pressed].starts_with?("vm_", # Handle buttons from sub-items screen
-                                     "miq_template_",
-                                     "guest_",
-                                     "host_")
+    return %w{
+      custom_button host_analyze_check_compliance host_check_compliance
+      host_delete host_edit host_protect host_refresh host_scan host_tag
+      storage_delete storage_refresh storage_scan storage_tag
+    }.include?(pressed)
```

Comment:
I don't like that here you have a second place, where you enumerate the buttons (first is the case statement at the start of this method). 

Can you use the when/else statement to handle the situation where none of the buttons is pressed? Or figure something better...